src/Nethermind/Nethermind.Core/ChainLevelInfo.cs

```diff
@@ -83,14 +83,17 @@ public void InsertBlockInfo(Keccak hash, BlockInfo blockInfo, bool setAsMain)
             BlockInfo[] blockInfos = BlockInfos;
 
             int? foundIndex = FindIndex(hash);
-            if (!foundIndex.HasValue)
+            if (foundIndex is null)
             {
                 Array.Resize(ref blockInfos, blockInfos.Length + 1);
             }
             else
             {
                 if (blockInfo.IsBeaconInfo && blockInfos[foundIndex.Value].IsBeaconMainChain)
                     blockInfo.Metadata |= BlockMetadata.BeaconMainChain;
+
+                if (blockInfo.EqualsIgnoringWasProcessed(blockInfos[foundIndex.Value]))
+                    blockInfo.WasProcessed |= blockInfos[foundIndex.Value].WasProcessed;
```

Comment:
Actual fix: when updating a `BlockInfo` only on the `WasProcessed` flag, prefer the truth value.

Example: Consider two `BlockInfo` `x` and `y` that are the same except for `WasProcessed` where `x.WasProcessed == true` and `y.WasProcessed == false`. If we want to update `x` with `y`, we do not override `x.WasProcessed` with `false` since it does not make sense: we already processed a block with the exact information contained in `x`. The opposite is not true: if we want to update `y` with `x` then we override the `x.WasProcessed` with `true` since we're stating that such block has now been processed.

Reply:
If the problem is only in genesis, then why not specify genesis as always processed when it recreate genesis?

Reply:
Good point, we could do this:

```c#
// src/Nethermind/Nethermind.Blockchain/BlockTree.cs
if (current.IsGenesis)
{
    current.TotalDifficulty = current.Difficulty;
    BlockInfo blockInfo = new(current.Hash, current.Difficulty)
    {
        WasProcessed = true,
    };
    UpdateOrCreateLevel(current.Number, current.Hash, blockInfo);
}
```

During testing, this also fixed the issue, but what about other blocks that are being updated to not processed while having the same information? Is it possible to break those?

Reply:
If a block was processed but we had consensus issue, then we updated the client, and now it's processed with different outcome, could it lead to any problem? A user may also update update fork parameters and restart I guess.

Reply:
Hm what if genesis config is changed btw?

Reply:
Is there any other scenario where this could go wrong?
I think we should just set genesis.WasProcessed to true when loading genesis, would be simpler.